common/diagnostics/MonitoringServer.java

```diff
@@ -0,0 +1,154 @@
+        private final ChannelInboundHandlerAdapter[] middleware;
+
+        MetricsInitializer(@Nullable SslContext sslContext, ChannelInboundHandlerAdapter... middleware) {
+            this.sslContext = sslContext;
+            this.middleware = middleware;
+        }
+
+        @Override
+        protected void initChannel(SocketChannel socketChannel) {
+            ChannelPipeline pipeline = socketChannel.pipeline();
+            if (sslContext != null) {
+                pipeline.addLast(sslContext.newHandler(socketChannel.alloc()));
+            }
+            pipeline.addLast(new HttpServerCodec());
+            pipeline.addLast(new HttpServerExpectContinueHandler());
+            for (ChannelInboundHandlerAdapter handler : middleware) {
+                pipeline.addLast(handler);
+            }
+            pipeline.addLast(new MetricsHandler());
+        }
+    }
+
+    class MetricsHandler extends SimpleChannelInboundHandler<HttpObject> {
+        @Override
+        public void channelRead0(ChannelHandlerContext ctx, HttpObject msg) {
+            if (msg instanceof HttpRequest) {
+                HttpRequest req = (HttpRequest) msg;
+                URI uri = URI.create(req.uri());
+
+                if (!uri.getPath().equals("/metrics")) {
+                    FullHttpResponse response = new DefaultFullHttpResponse(req.protocolVersion(), NOT_FOUND);
+                    response.headers().set(CONTENT_TYPE, TEXT_PLAIN).setInt(CONTENT_LENGTH, response.content().readableBytes()).set(CONNECTION, CLOSE);
+                    ctx.write(response);
+                    return;
+                }
+
+                FullHttpResponse response = new DefaultFullHttpResponse(req.protocolVersion(), OK);
+                String query = Objects.requireNonNullElse(uri.getQuery(), "");
+                if (query.toLowerCase().contains("format=json")) {
+                    response.content().writeBytes(Unpooled.wrappedBuffer((metrics.formatJSON()).getBytes(StandardCharsets.UTF_8)));
+                    response.headers().set(CONTENT_TYPE, APPLICATION_JSON);
+                } else {
+                    response.content().writeBytes(Unpooled.wrappedBuffer((metrics.formatPrometheus()).getBytes(StandardCharsets.UTF_8)));
+                    response.headers().set(CONTENT_TYPE, TEXT_PLAIN);
+                }
+                response.headers().setInt(CONTENT_LENGTH, response.content().readableBytes()).set(CONNECTION, CLOSE);
+
+                ctx.write(response).addListener(ChannelFutureListener.CLOSE);
+            }
+        }
+
+        @Override
+        public void channelReadComplete(ChannelHandlerContext ctx) {
+            ctx.flush();
+        }
+
+        @Override
+        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Failed to respond to a metrics request:\n{}\n", Arrays.toString(cause.getStackTrace()));
```

Comment:
I believe the LOGger can take an exception as an argument and will handle formatting and stuff for you, check some other places we log errors